WIP: Deterministic encryption for ansible-vault (Implementing #35480)

agowa opened the pull request:
##### SUMMARY
The salt that is used for generating the keys and the iv's is now derived from the plaintext and the secret.
Therefore the same key and plaintext generates always the same ciphertext.

Also a minore refactoring is done to simplify implementing of future encryption algorithms.
The encryption whitelist and write_whitelist is now generated dynamically.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
Vault

##### ADDITIONAL INFORMATION
Before the salt was derived by calling `os.urandom(32)` which provides 256 random bits.
Now it is derived by:
sha256(b_plaintext + b_secret)
and if either b_plaintext or b_secret is None (what should never happen, except code reuses) the salt is still calculated randomly, as it could simplify brute forcing the plaintext or the passphrase, as an attacker could simply do:
```
for e in dictionary:
  if sha256(e) == sha256(b_plaintext):
    print(e)
    break
```
and if parallel execution of this loop is considered (rewritten for parallel execution), it gets even worse, if you choose weak input. To prevent this, the plaintext and the secret are concatenated, as the secret is said to be secure (and if it is not, we really have other problems) and the plaintext is unknown to the attacker. So the only way an attacker has to brut force is is by trying to decrypt.

Fixes #35480

jamescassell commented:
Likewise, I'd feel better if a crypto expert looked at it first.

agowa338 commented:
jborean93: Basically what I've said in the introduction, having the same ciphertext for the same plaintext.

The salt does not need to be random, it needs to be unique, as it is public knowledge, it is stored in plaintext in the resulting file. So it does not matter how it is generated, as the attacker can always read it.

There are side effect of making it dependent on the plaintext, as stated above, it will leak the information, that the file is identical to another file, as same plaintext ensures same ciphertext. But as those files are checked into version control, this is wanted. Also currently you simple did not re encrypt the encrypted files with each commit, so an attacker also knows that those file/files did not change or that they changed.

And the second one is one has to look at how the determinism works. In this commit I concatenate the plaintext and the secret and generate a sha256 sum of them. This sha256 sum is stored unencrypted (as the salt) with the ciphertext. This is something that does not provide any value to attackers, as the problem to derive the plaintext of the function `sha256(plaintext+secret)` can be considered at all times at least as hard as brute forcing the secret by `aes256^-1(ciphertext, salt, secret)`. And this problem is the same as currently.

The only time an attacker can solve `sha256^-1(ciphertext)` for `plaintext+secret`, is if it is already known to the attacker (or known enough to be brut forced), wich is equal to having a insecure secret. But than the problem `aes256^-1(ciphertext, salt, secret)` is easier to solve than `sha256^-1(ciphertext) => plaintext+secret`.

Also sha256 is a irreversible cryptographic secure hashing function, so a straight calculation from the cipher text to the plain is not possible and flipping a bit in the plaintext will result in unpredictable different output.

Basically a secure (https://en.wikipedia.org/wiki/Convergent_encryption)[convergent encryption] is what one would want to use in combination with version control. The simplest method of archiving this is by switching AES-CTR with AES-ECB, but aes-ecb would leak way to much information from the resulting ciphertext, as multiple inputs share the same key and therefore lack diffusion. A much better way is to use [https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_(CTR)](AES-CTR) with a deterministic nonce. And as the nonce (and the counter) is required to decrypt the ciphertext, it is not that relevant to be random, but only to be unique for different files.

At a later time more aes modes could be implemented, like AES-OFB (which allows better error correction), but also propagates a bitflip in the ciphertext back to the plaintext. Or RC6, Twofish, MARS, Serpent.

jamescassell: Yes, I'm appreciating a review, as this is a security related change.

jamescassell commented:
Could you remove that leakage by also basing the salt on the file path/name and the last commit hash prior to the one that last changed the file?

agowa338 commented:
jamescassell: Using the last commit hash would not provide the same ciphertext if the file is de/encrypted. Can you please be more specific on what leakage you mean? Am I correct to assume you mean the file a is equal to file b one? If so, having this information 'leaked' is the whole point of this change.

jborean93 commented:
Thanks for the info @agowa338 this is definitely a field I'm interested in but not really too knowledgeable. For the changes, I recommend you raise an item in the meeting agenda and talk about this at the next meeting to get a general consensus from the core team. The agenda list is here https://github.com/ansible/community/issues/329 and the scheduled is here https://github.com/ansible/community/blob/master/meetings/README.md.

bcoca commented:
by changing to a blacklist it seems we now allow TEST as a cypher ...

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -1356,11 +1354,51 @@ def _encrypt_pycrypto(b_plaintext, b_key1, b_key2, b_iv):
+        # sha256 hash of the secret.
+
+        if not secret or not b_plaintext:
+            return os.urandom(32)
```
Since we're falling back to os.urandom in the case of empty, perhaps we should also fall back in the case of not enough entropy in the plaintext?  Is something like len(b_plaintext ) < 256 sufficient for that?

Should a similar check be applied to the secret?

agowa replied:
We could do that, but it wouldn't provide any additional security, it would even weaken it, as an attacker could guess that the file was not changed and therefore knows that it has weak content or secret as it is checked in with every commit and can be observed to change too frequently.
Having a length check here, could potentially leak the fact that the file is a good target for an attack. And we want to protect the user from themselves, so we need to not threat it differently and recommend using long and complex passphrases (like we hopefully already do)

abadger replied:
<nod>  I see what you mean about checking the length for the secret leaking information about the secret that the attacker would not otherwise know.  However, checking the length for the plaintext doesn't seem to suffer from that same problem.  The attacker already has an idea of how long the plaintext is because the length of the ciphertext will reflect that.  So checking the length of the plaintext and falling back to os.urandom() if the plaintext is too short seems like it can prevent some problems without revealing new information?  (Hmm... it would reveal that the plaintext is less than a precise value whereas looking at the length of the ciphertext  would be more vague.... Would that be important or not?)

The attacker can learn a lot about what is in a vault file from external sources... for instance, since they're yaml files, they likely have a rough outline of:
```
---
var: value
```

an attacker might also find the precise name to use for "var" by looking at playbooks to see which variable names exist that are not defined outside of a playbook.  So then the only thing about the plaintext that they'd be lacking would be the actual value....

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -1473,3 +1511,14 @@ def decrypt(cls, b_vaulttext, secret):
+CIPHER_WRITE_WHITELIST = frozenset(
+    [item for item in CIPHER_MAPPING.keys() if item not in CIPHER_WRITE_BLACKLIST]
+)
+# Also see DEFAULT_CIPHERS at the top of this file
```
We should leave the blacklists and whitelists alone for now.  I don't like adding unnecessary symbols (So no blacklists, for instance).  Everything should be a frozenset.  Generating the whitelists from CIPHER_MAPPING.keys() does have some appeal but considering that there's currently only two ciphers, not very much compared to creating the blacklist unneccessarily.

Those difficulties could all be hashed out in a different PR if we want.

agowa replied:
ok, I'll remove it for now.

agowa replied:
ok, reverted the change.

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -1356,11 +1354,51 @@ def _encrypt_pycrypto(b_plaintext, b_key1, b_key2, b_iv):
+        # If we would do that, we leak the sha256 hash of the plaintext.
+        # This is most of the time not a problem, except if the plaintext contains guessable
+        # content or is partially known by the attacker. The attacker then could simply
+        # broothforce the plaintext content. This is not a problem for larger files, where
```
s/broothforce/bruteforce/g

agowa replied:
done

agowa replied:
done

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -625,9 +626,6 @@ def encrypt(self, plaintext, secret=None, vault_id=None):
             raise AnsibleError("input is already encrypted")
 
-        if not self.cipher_name or self.cipher_name not in CIPHER_WRITE_WHITELIST:
-            self.cipher_name = u"AES256"
```
We do want to keep the CIPHER_WRITE_WHITELIST check here for sanity.  It could be changed to throw an error, though, as cipher_name should be set in __init__ with your change.

agowa replied:
Re added the if and changed to throw an exception

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -204,7 +203,7 @@ def parse_vaulttext_envelope(b_vaulttext_envelope, default_vault_id=None, filena
 
 
-def format_vaulttext_envelope(b_ciphertext, cipher_name, version=None, vault_id=None):
+def format_vaulttext_envelope(b_ciphertext, cipher_name=DEFAULT_CIPHER, version='1.1', vault_id=None):
```
I think we should leave the cipher_name as mandatory.  The cipher is associated with the ciphertext and should become mandatory at the same time.  (You don't want a plaintext to be enciphered with Algorithm Foo but the cipher_name to be "Algorithm Bar").  AFAICS, your change to set the version in the parameter list is fine, though.

abadger replied:
well, actually, I guess I would leave it None in the parameter list and change the setting of version which happens below to take the vault_id special case into account like this:
``` python
if vault_id and vault_id != u'default':
    if not version:
        version = '1.2'
    else:
        raise AnsibleError("The vault format version must be at least 1.2 if vault_id is specified")
elif not version:
    version = '1.1'
```

@alikins wrote the vault_id stuff, though, so he would be the one to ask if there's a reason he coded this the other way.

(Note that this also appears to be a tangent to the main purpose of the PR, so perhaps it should be spearated into its own PR).

agowa replied:
done

abadger commented:
I've reviewed the things I can here.  We would want to get this passed through a security review before accepting it, though.  We'll see what the rest of the team thinks at the meeting tomorrow and if no one has objections we'll have to seek the security review externally from our team.

agowa committed:
Make the ansible-vault encryption deterministic to allow better handling inside a version controlled environment.

This leaks the information that a file is identical to another (sha256 hash).
Simply ciphertext stays the same if the plaintext is the same.
As ansible-vault encrypted files are supposed to be version controlled, this information is already public knowledge.

Don't leak sha256 hash of plaintext file anymore by salting it with the secret.
So now we salt the salt with the secret.

Implementing the feature request #35480

abadger commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -1356,11 +1355,51 @@ def _encrypt_pycrypto(b_plaintext, b_key1, b_key2, b_iv):
+            digest.update(b_plaintextAndSecret)
+            b_salt = digest.finalize()
+        elif HAS_PYCRYPTO:
+            b_salt = SHA256_pycrypto.new(b_plaintextAndSecret).digest()
```
In today's meeting, @jborean93 brought up that we could use a pbkdf function here to generate the salt rather than a simple sha256sum hash.  The rationale would be if one was attempting to brute force the hash by guessing the plaintext and secret, they would have to run their guess through more hashing rounds if we use a pbkdf function than if we simply run sha256sum on the input once.

To use a pbkdf function we'd have to hardcode a salt to use with that function but I don't think that's any worse than calling sha256sum() on  b_plaintextAndSecret with no salt applied.  Correct?

abadger replied:
For context, this was proposed because the difference between trying to brute force the ciphertext and trying to brute-force the hash being used for the salt was brought up.  The pbkdf function making it more expensive to generate the encryption key from the password rather than generating the hash was one difference that was brought up.

nitzmahone commented:
I'm a pretty strong -1 for this. Deterministic encryption is generally considered a bug, not a feature, in a system like this. Realistically, the duplicate-detection mechanism you're trying to achieve probably won't work very well anyway in a real-world vault store where multiple people are making changes, reordering keys, introducing extra whitespace, etc.

You might want to explore implementing an editor shim to perform custom logic/hashing/canonicalization out-of-band on the plaintext, rather than relying on comparisons of arbitrary ciphertexts to detect the conditions you're looking for. The only immediate impediment I can think of there would be that we currently ignore the exit code from the editor, but that could be easily rectified (to allow your editor shim to fail the encryption if your rules were violated).

agowa338 commented:
> Realistically, the duplicate-detection mechanism you're trying to achieve probably won't work very well anyway in a real-world vault store where multiple people are making changes, reordering keys, introducing extra whitespace, etc.

You miss, that there is also inline encryption of ansible vault, where you only encrypting a single item. And also if one decrypts all files within a project after check out and re encryptes prior to commiting the changes, the git history gets clumbered with "File x changed" even though it did not. That's the problems I want to solve. By using a deterministic encryption, these files are recognized by git as not changed and therefore will not get staged.
If the inline encryption is used, it also allows one to do regression tracking, you changed the database password, and your college changes it back to it's old value.

Using an editor shim won't work for diffing in the web ui, like github or gitlab. Also if you use git diff, smudge and clean filters, it does not disaper from "git status", as it compares the resulting files. So it will happen, that "git diff" shows nothing but "git status" every file that is ansible vault encrypted.

What I've considered is having this deterministic encryption as another algorithm that can be chosen by the user, but that would require implementing an additional command line switch. And if such a command line parameter is added, it should be generic enough to allow implementing additional algorithms in the future.
Like "ansible-vault encrypt --cipher aes --mode ci --salt sha256 file.vault.yml" bzw. with this oob defaults: "--cipher aes --mode ci --salt random". Same for encrypt_string.

Also another alternative if using sha(plaintextAndSecret) is considered not good enough (despite it being secure or not) on could implement AES-SIV ([https://tools.ietf.org/html/rfc5297](rfc5297), [https://github.com/miscreant/miscreant/wiki/AES-SIV](Graphic)). But that also would require implementing the above mentioned command line parameters and also an additional value for the cipher-header-filed.

nitzmahone commented:
My primary concern is that changing the default behavior in this way provides some expectation of determinism forever; that potentially ties our hands when selecting the next vault encryption scheme. I'd be less worried about it if it were a config/cmdline switch where the user is explicitly opting into a specific cipher/salting scheme (I suspect tricky to implement in a foolproof way)... But a hypothetical generic `--deterministic` option implies that we're always going to provide a secure deterministic option, which may not always be possible.

tiran commented:
@agowa338, your initial message describes **what** you want to change, but not **why** you want to change the encryption scheme. Could you be so kind and explain, which problem your patch is trying to solve and why you consider the current indeterministic encryption scheme as inferior to your proposed deterministic encryption scheme.

agowa338 commented:
two comments above yours:
> And also if one decrypts all files within a project after check out and re encryptes prior to commiting the changes, the git history gets clumbered with "File x changed" even though it did not. That's the problems I want to solve. By using a deterministic encryption, these files are recognized by git as not changed and therefore will not get staged.
If the inline encryption is used, it also allows one to do regression tracking, you changed the database password, and your college changes it back to it's old value.

agowa committed:
Harden against cpa

agowa338 commented:
I flipped the order of concatenation, so that the CPA does not compromise the key, as an attacker could calculate the sha256 to the most common ancestor and start brute forcing from there (reducing the time and calculations necessary). Now by flipping the order an attacker cannot start from this known sha256 inner state and is forced to do the whole sha256 calculation over and over again (except the secret is known by the attacker, but brute forcing the plaintext does not have practical use if one already knows the secret). For the same reason I added another round of sha256, which will also hide the true sha256 of the secret and the plaintext, as reversing the outer layer of sha256 will (if it should be at some point in time) have many hash collisions and an attacker would have to crack every one of those hashes until he finds the secret. Effectively this implements a much better hardening against CPA attacks, as outlined above.

I consider the scenario 'please encrypt "hello1234" as I suspect it is your password' kind of attack as out of scope, as it is basically what we want to have. We want to know if a given plaintext matches a given ciphertext.
So basically:
Plaintext_a == Plaintext_b   =>   Ciphertext_a == Ciphertext_b
And therefore I don't consider this an attack, but a feature.

melbit-naveennathan commented:
I would definitely like to see deterministic encryption but of AES-SIV variety (e.g. IV=MAC(P), C'Enc(IV, K, P), C=MAC(P)||C'). However I believe to achieve proper security you still need to come up with a strong cryptographic key and I'm not sure what security drawbacks there are with a deterministic salt. If using a randomised salt, then it would be pointless to use AES-SIV, since the cipher will always be random per encryption.

agowa338 commented:
@melbit-naveennathan AES-SIV has a deterministic salt, that's the difference to AES-CTR ;-) I'll reevaluate that, but as discussed elsewhere, that's possibly the best option, as it also provides authentication of encrypted messages.

melbit-naveennathan commented:
No AES-SIV has a deterministic _IV_. The salt is an input to PBKDF2. Combining PBKDF2 using randomised salt to produce encryption/mac keys for AES-SIV is somewhat pointless because the ciphertext will be randomised. The salt needs to be unpredictable and deterministic but shouldn't leak any information about the plaintext or the password input of PBKDF2.

For example a salt using your original proposal where `sha256(b_plaintext + b_secret)` (where b_secret is the password input to PBKDF2), this is inadvisable . If the attacker knows the plaintext (and this is usually the case when you consider threat models such as IND-CPA, IND-CCA2, etc.), it can launch a dictionary attack on the salt by computing the _fast_ hash `sha256(known_plaintext + guess)` until it matches the salt, negating the protections that PBKDF2 provide.

Deterministic encryption would be nice if it can be implemented properly (as in SIV mode) but doesn't seem to interoperate well with PBKDF2.

agowa338 commented:
As I said, I'll re-evaluate that.

agowa338 commented:
@melbit-naveennathan Ok, I've taken a quick look and just want to get you opinion on it. This is not yet final, but should be secure™, but I still need to check that.

> I would definitely like to see deterministic encryption but of AES-SIV variety (e.g. IV=MAC(P), C'=Enc(IV, K, P), C=MAC(P)||C'). However I believe to achieve proper security you still need to come up with a strong cryptographic key and I'm not sure what security drawbacks there are with a deterministic salt. If using a randomised salt, then it would be pointless to use AES-SIV, since the cipher will always be random per encryption.

KDF:
* PBKDF2 can use hmac.new("key", "plaintext", hashlib.sha3_256) as salt
* Argon2 also with hmac.new("key", "plaintext", hashlib.sha3_256) as salt

Cipher:
* Either stick to AES-CTR and just use the new KDF
* Or switch to AES-SIV `Crypto.Cipher.AES.new(key, AES.MODE_SIV, *, nonce=None)`, for aes256, key needs to be 512 bits long (aka. the hmac will be sha3_512 instead of sha3_256).

Using Argon2 and AES-SIV should provide message authentication as well as deterministic encryption without leaking information despite two files being identical.

melbit-naveennathan commented:
> * PBKDF2 can use hmac.new("key", "plaintext", hashlib.sha3_256) as salt
> * Argon2 also with hmac.new("key", "plaintext", hashlib.sha3_256) as salt

Assuming "key" in this context to be the input password, an attacker that knows the plaintext can subvert the password hash by simply performing a dictionary attack using faster HMAC, negating the security properties of the password hash.

This should be enough to avoid the construction in the first place, but

> * Either stick to AES-CTR and just use the new KDF
I'm not entirely sure about this scheme since this is probably not secure "authenticated encryption" scheme and instead appears to be more of the MAC-then-encrypt variety.

> * Or switch to AES-SIV `Crypto.Cipher.AES.new(key, AES.MODE_SIV, *, nonce=None)`, for aes256, key needs to be 512 bits long (aka. the hmac will be sha3_512 instead of sha3_256).
Both PBKDF2 and Argon2 should be able to extract 512 bits so you can split the output for encryption and MAC key. I'm still unsure about the syntax of the encryption since the nonce I assume is the IV that is generated using the HMAC(k,plaintext) in this case?


The construction I considered is:
salt=0
ek,mk=PBKDF2(salt, password) # or argon2/scrypt
IV=HMAC(mk, plaintext)
output: IV,AES-256-CTR(ek,plaintext)

The problem with the above construction is that even though it uses a provably secure deterministic authentication encryption scheme, the password hash is weakened because it uses a static salt, so an attacker could perform a rainbow table attack which is much faster than enumerating a dictionary and computing the password hash each time.

I've come to the conclusion that a deterministic authenticated encryption scheme simply cannot operate securely with a password hash without weakening the password hash construct.

agowa338 commented:
@melbit-naveennathan I think I found the ideal solution. Sometimes going back a step can lead you forward. The original problem is to have the vault encrypted data in plaintext within the working directory, but still uploading the ciphertext.
Therefore storing the salt when decrypting a file and reusing it when re-encrypting it (as long as it's hash/hmac has not changed) should be sufficient. And in all other cases use a random salt.

Therefore for the first invocation a random salt will be generated. If the file did not change, it will be re-encrypted with the same salt, resulting in the same ciphertext, allowing `git status` to show the file as not changed and a git filter to diff it's plaintext. Than a git hook can be used for transparent en-/decryption.
If the content of the file changes, the cached hmac salt tuple does no longer match and therefore a new random salt is generated. The cached data is blacklisted using .gitignore and therefore it is never committed into the repo.

And using argon2 instead of PBKDF2 and using AES-SIV instead of AES-CTR will increase security, but they could be implemented as alternatives instead of replacing the current one, maybe becoming the default.

So what do you think?

melbit-naveennathan commented:
Sure, you could expose salt and IV parameters to ansible-vault which will let you control the randomness of the encryption scheme.

s-hertel commented:
We're interested in supporting this via the approach in #71424. A more invasive change like this really needs a review from someone with a cryptography background.
